### PR TITLE
feat: Add stiffness matrix computation in integrators

### PR DESCRIPTION
**Pull request by talledodiego**
This PR adds to the integrators the computation of section stiffness matrix.

The steps are:

- [x] Refactor fiber integrator simplifying the function prepare_input factoring out a triangulate function
- [x] Refactor marin integrator simplifying the function prepare_input
- [x] Rename prepare_input, integrate and integrate_strain_response_on_geometry indicating that those methods are for integration of stresses (for computing section forces)
- [x] Add the proper methods for computing stiffness matrix for fiber integrator
- [x] Add marin coefficients computation for integrating the stiffness
- [x] Add the proper methods for computing stiffness matrix for marin algorithm
- [x] Add relevant tests

### CODE REVIEW

File: structuralcodes/sections/_generic.py [hunk 1]
```diff
@@ -196,7 +196,7 @@ def compute_area_moments(geometry, material=None):
                 iyy,
                 iyz,
                 tri,
-            ) = self.integrator.integrate_strain_response_on_geometry(
+            ) = self.integrator.integrate_strain_response_on_geometry_stress(
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:199]: Perhaps the name is starting to become a bit long? What if we simply call the method `stress_resultant`?

File: structuralcodes/sections/section_integrators/_fiber_integrator.py [hunk 2]
```diff
@@ -228,9 +249,115 @@ def integrate_strain_response_on_geometry(
...
+            geo, strain, **kwargs
+        )
+
+        # Return the calculated response
+        return *self.integrate_stress(prepared_input), triangulated_data
+
+    def prepare_input_tangent(
+        self, geo: CompoundGeometry, strain: ArrayLike, **kwargs
+    ) -> t.Tuple[t.Tuple[np.ndarray, np.ndarray, np.ndarray]]:
+        """Prepare general input to the integration of stresses in the section.
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_fiber_integrator.py:262]: We should also update the docstring here.

File: structuralcodes/sections/section_integrators/_fiber_integrator.py [hunk 3]
```diff
@@ -228,9 +249,115 @@ def integrate_strain_response_on_geometry(
...
         # Prepare the general input based on the geometry and the input strains
-        prepared_input, triangulated_data = self.prepare_input(
+        prepared_input, triangulated_data = self.prepare_input_stress(
+            geo, strain, **kwargs
+        )
+
+        # Return the calculated response
+        return *self.integrate_stress(prepared_input), triangulated_data
+
+    def prepare_input_tangent(
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_fiber_integrator.py:259]: To this and the other related methods. Do we anticipate that we will only include the tangent stiffness, and not a secant or initial etc.? If we anticipate to expand later, I think we should avoid calling the method `_tangent`.

File: structuralcodes/sections/section_integrators/_fiber_integrator.py [hunk 4]
```diff
@@ -228,9 +249,115 @@ def integrate_strain_response_on_geometry(
...
+        tangent[0, 0] = np.sum(TA)
+        tangent[0, 1] = tangent[1, 0] = np.sum(TA * y)
+        tangent[0, 2] = tangent[2, 0] = np.sum(-TA * x)
+        tangent[1, 1] = np.sum(y * y * TA)
+        tangent[1, 2] = tangent[2, 1] = np.sum(-x * y * TA)
+        tangent[2, 2] = np.sum(x * x * TA)
+
+        return tangent
+
+    def integrate_strain_response_on_geometry_tangent(
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_fiber_integrator.py:339]: Same as above, the name starts to become a bit long. Should we consider `section_stiffness` or something? What do you think?

**Review by mortenengen — COMMENTED** [review]
Good job so far on this one. I have left a couple of comments for you to consider.

**Comment by talledodiego** [comment]
@mortenengen ready for your review.

File: structuralcodes/sections/_generic.py [hunk 5]
```diff
@@ -604,25 +606,49 @@ def _rotate_triangulated_data(self, theta: float):
         self.triangulated_data = rotated_triangulated_data
 
     def integrate_strain_profile(
-        self, strain: ArrayLike
-    ) -> t.Tuple[float, float, float]:
+        self,
+        strain: ArrayLike,
+        integrate: t.Literal['stress', 'modulus'] = 'stress',
+    ) -> t.Union[t.Tuple[float, float, float], NDArray]:
         """Integrate a strain profile returning internal forces.
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:613]: Suggest to change to: ...returning internal forces or section stiffness.

File: structuralcodes/sections/_generic.py [hunk 6]
```diff
@@ -604,25 +606,49 @@ def _rotate_triangulated_data(self, theta: float):
...
             strain=strain,
+            integrate=integrate,
             tri=self.triangulated_data,
             mesh_size=self.mesh_size,
         )
-        return N, My, Mz
+
+        if len(result) == 2:
+            return result[0]
+        return result[:-1]
```
Comment (mortenengen) [inline structuralcodes/sections/_generic.py:651]: Please consider adding a comment explaining the purpose of these lines.

File: structuralcodes/sections/section_integrators/_fiber_integrator.py [hunk 7]
```diff
@@ -60,26 +61,135 @@ def prepare_triangulation(self, geo: SurfaceGeometry) -> t.Dict:
...
+        geo: CompoundGeometry,
+        strain: ArrayLike,
+        integrate: t.Literal['stress', 'modulus'] = 'stress',
+        **kwargs,
     ) -> t.Tuple[t.Tuple[np.ndarray, np.ndarray, np.ndarray]]:
-        """Prepare general input to the integration.
+        """Prepare general input to the integration of stresses in the section.
 
-        Calculate the stresses based on strains in a set of points.
+        Calculate the forces based on strains in a set of points.
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_fiber_integrator.py:167]: Suggest to change the signature to e.g.:

```
    ... integration of stress or material modulus in the section.
    ... the stress resultants or tangent section stiffness matrix based on ...
```

File: structuralcodes/sections/section_integrators/_fiber_integrator.py [hunk 8]
```diff
@@ -202,35 +241,89 @@ def integrate(
...
-        self, geo: CompoundGeometry, strain: ArrayLike, **kwargs
-    ):
-        """Integrate the strain response with the fiber algorithm.
+        self,
+        geo: CompoundGeometry,
+        strain: ArrayLike,
+        integrate: t.Literal['stress', 'modulus'] = 'stress',
+        **kwargs,
+    ) -> t.Tuple[t.Union[t.Tuple[float, float, float], np.ndarray], t.List]:
+        """Integrate stresses the strain response with the fiber algorithm.
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_fiber_integrator.py:287]: Suggest to change the first line in the docstring in a similar manner as above to also describe integrating modulus.

File: structuralcodes/sections/section_integrators/_marin_integrator.py [hunk 9]
```diff
@@ -126,29 +132,92 @@ def get_input_polygon(polygon, coeffs):
...
+
+    def prepare_input(
+        self,
+        geo: CompoundGeometry,
+        strain: ArrayLike,
+        integrate: t.Literal['stress', 'modulus'] = 'stress',
+    ) -> t.Tuple[float, t.Tuple[np.ndarray, np.ndarray, np.ndarray]]:
+        """Prepare general input to the stress integration.
+
+        Calculate the stresses based on strains in a set of points.
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_marin_integrator.py:175]: Suggest to modify the docstring similar to the suggestions above.

File: structuralcodes/sections/section_integrators/_marin_integrator.py [hunk 10]
```diff
@@ -176,47 +245,143 @@ def integrate(
...
     def integrate_strain_response_on_geometry(
-        self, geo: CompoundGeometry, strain: ArrayLike, **kwargs
+        self,
+        geo: CompoundGeometry,
+        strain: ArrayLike,
+        integrate: t.Literal['stress', 'modulus'] = 'stress',
+        **kwargs,
     ):
-        """Integrate the strain response with the Marin algorithm.
+        """Integrate the strees from strain response with the Marin algorithm.
```
Comment (mortenengen) [inline structuralcodes/sections/section_integrators/_marin_integrator.py:350]: Suggest to modify the docstring similar to the suggestions above.

File: tests/test_core/test_generic_section.py [hunk 11]
```diff
@@ -128,6 +128,382 @@ def test_rectangular_section():
...
+    assert sec.name == 'GenericSection'
+
+    assert math.isclose(sec.gross_properties.area, b * h)
+
+    # compute stiffness matrix
+    stiffness = sec.section_calculator.integrate_strain_profile(
+        [0, 0, 0], 'modulus'
+    )
+    assert stiffness.shape == (3, 3)
+    stiffness /= E
```
Comment (mortenengen) [inline tests/test_core/test_generic_section.py:179]: Could this part be even stronger if we parameterize the name of the integrator, i.e. `'fiber', 'marin'`, and check both against the analytical solution?

File: tests/test_core/test_generic_section.py [hunk 12]
```diff
@@ -128,6 +128,382 @@ def test_rectangular_section():
...
+    EIzz_gross = Ec * 1 / 12.0 * h * b**3
+    for p in geo.point_geometries:
+        y, z = p.point.coords.xy
+        As = p.area
+        EA_gross += Es * As
+        EIyy_gross += Es * As * z[0] ** 2
+        EIzz_gross += Es * As * y[0] ** 2
+
+    # Create the section with fiber integrator
+    sec = GenericSection(geo, integrator='fiber', mesh_size=0.0001)
```
Comment (mortenengen) [inline tests/test_core/test_generic_section.py:253]: Same here, could we parameterize the name of the integrator?

**Review by mortenengen — CHANGES_REQUESTED** [review]
Great job on this one! I have left a couple of minor comments in the code.

There are some lines that are still not covered by the tests. For the integrators, the lines are "nice-to-cover", i.e. the missing lines are mostly exceptions. However, there are also quite a number of "need-to-cover" lines related to the constitutive laws that are not covered. I know that many of these lines are not treated in this PR, but I think that we should at least cover the ones related to the Marin coefficients. What do you think?

File: structuralcodes/sections/_generic.py (hunk 5 above)
Comment (talledodiego) [reply structuralcodes/sections/_generic.py:613]: sure, thanks!

File: structuralcodes/sections/_generic.py (hunk 6 above)
Comment (talledodiego) [reply structuralcodes/sections/_generic.py:651]: done

File: structuralcodes/sections/section_integrators/_fiber_integrator.py (hunk 7 above)
Comment (talledodiego) [reply structuralcodes/sections/section_integrators/_fiber_integrator.py:167]: done

File: structuralcodes/sections/section_integrators/_fiber_integrator.py (hunk 8 above)
Comment (talledodiego) [reply structuralcodes/sections/section_integrators/_fiber_integrator.py:287]: done

File: structuralcodes/sections/section_integrators/_marin_integrator.py (hunk 9 above)
Comment (talledodiego) [reply structuralcodes/sections/section_integrators/_marin_integrator.py:175]: done

File: structuralcodes/sections/section_integrators/_marin_integrator.py (hunk 10 above)
Comment (talledodiego) [reply structuralcodes/sections/section_integrators/_marin_integrator.py:350]: done

File: tests/test_core/test_generic_section.py (hunk 11 above)
Comment (talledodiego) [reply tests/test_core/test_generic_section.py:179]: sure!

File: tests/test_core/test_generic_section.py (hunk 12 above)
Comment (talledodiego) [reply tests/test_core/test_generic_section.py:253]: sure!

**Comment by talledodiego** [comment]
I included much more tests for marin coefficients! Can this be considered enough?

**Review by mortenengen — APPROVED** [review]

**Pull request merged**